File: 203.移除链表元素/removeElements.go
package problem0203

type ListNode struct {
	Val  int
	Next *ListNode
}

// 要删除当前节点，需要记录上一个节点，因此需要一个prev节点，并且初始状态prev.Next = head
// 最后返回prev.Next
func removeElements(head *ListNode, val int) *ListNode {
	dummy := &ListNode{Next: head}
	prev := dummy

	for head != nil {
		if head.Val == val {
			prev.Next = head.Next
		} else {
			prev = head
		}
		head = head.Next
	}

	return dummy.Next
}

File: 203.移除链表元素/removeElements.js
// 移除链表元素
// 给你一个链表的头节点 head 和一个整数 val ，请你删除链表中所有满足 Node.val == val 的节点，并返回 新的头节点 。
// 列表中的节点在范围 [0, 104] 内

function removeElements(head, val) {
	if (!head) {
		return head
	}
	const dummy = { next: head }
	let prev = dummy

	while (head) {
		if (head.val === val) {
			prev.next = head.next
		} else {
			prev = head
		}
		head = head.next
	}

	return dummy.next
}